Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp

```diff
@@ -28,26 +28,144 @@
 
 #include "GPUAdapter.h"
 #include "GPUCanvasConfiguration.h"
+#include "RenderBox.h"
+#if PLATFORM(COCOA)
+#include "WebGPUSurfaceCocoa.h"
+#endif
+#include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
+static IntSize getCanvasSizeAsIntSize(const GPUCanvasContext::CanvasType& canvas)
+{
+    return WTF::switchOn(canvas, [](const RefPtr<HTMLCanvasElement>& htmlCanvas) -> IntSize {
+        return { static_cast<int>(htmlCanvas->width()), static_cast<int>(htmlCanvas->height()) };
+    }
+#if ENABLE(OFFSCREEN_CANVAS)
+    , [](const RefPtr<OffscreenCanvas>& offscreenCanvas) -> IntSize {
+        return { static_cast<int>(offscreenCanvas->width()), static_cast<int>(offscreenCanvas->height()) };
+    }
+#endif
+    );
+}
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(GPUCanvasContext);
+
+std::unique_ptr<GPUCanvasContext> GPUCanvasContext::create(CanvasBase& canvas)
+{
+    return std::unique_ptr<GPUCanvasContext>(new GPUCanvasContext(canvas));
+}
+
+GPUCanvasContext::GPUCanvasContext(CanvasBase& canvas)
+    : GPUBasedCanvasRenderingContext(canvas)
+    , m_currentConfiguration()
+    , m_surface()
+{
+    suspendIfNeeded();
+}
+
+void GPUCanvasContext::reshape(int width, int height)
+{
+    if (!m_currentConfiguration)
+        return;
+
+#if PLATFORM(COCOA)
+    m_surface = WebGPUSurfaceCocoa::create(width, height, m_currentConfiguration->convertToBacking());
+    ASSERT(m_surface);
+#else
+    UNUSED_PARAM(width);
+    UNUSED_PARAM(height);
+#endif
+}
+
 GPUCanvasContext::CanvasType GPUCanvasContext::canvas()
 {
-    return { };
+    return htmlCanvas();
 }
 
-void GPUCanvasContext::configure(const GPUCanvasConfiguration&)
+void GPUCanvasContext::configure(GPUCanvasConfiguration&& configuration)
```

Comment:
You can make this `GPUCanvasContext::configure(GPUCanvasConfiguration configuration)` to support both a move and a copy if needed by implementors.

Reply:
Fixed.